### DISCUSSION TITLE
Question about how to use blImageReadFromFile/blImageWriteToFile

### DISCUSSION THREAD

**Discussion by f-saez** [blend2d/blend2d#208] [Q&A] [answered]
Hi,

I'm trying to use these two functions from another language than C++ and it seems I'm stuck.

The C++ examples use 
``` 
readFromFile(const char* fileName)
``` 

but with the C interface it more complicated as it is :
```
blImageReadFromFile(BLImageCore* self, const char* fileName, const BLArrayCore* codecs)
```

I've try to create a BLArrayCore of codecs but it doesn't work as my array is full of "empty" codecs.
So I've try use a null pointer for codecs.

It doesn't work and I've got an undocumented error message (2).
0 mean success but as the error messages start at 0x00010000, I've got no idea what 2 means.

**Comment by kobalicek** [comment]
Hi, can you describe where are you stuck at the moment?

This is as simple as it can get:

```
BLImageCore img;
const char* fileName = "some_image.png";

blImageInit(&img);
blImageReadFromFile(&img, fileName, NULL); // codecs can be null to use a built-in codecs list.
```

**Reply by f-saez** [reply]
thanks for the fast answer, but it is what I've tried and it doesn't work.

**Reply by kobalicek** [reply]
But that's not a sufficient information to help you. 

You can try to paste this code to the fiddle:

```c++
BLImage render(const BLContextCreateInfo& cci) {
  const char* fileName = "textures/tex00.jpg";

  BLImageCore img;
  BLResult result;

  blImageInit(&img);
  result = blImageReadFromFile(&img, fileName, nullptr);

  if (result != BL_SUCCESS) {
    printf("Failed to load %s\n", fileName);
  }

  return img.dcast();
}
```

The only C++ piece is `img.dcast()` which takes care of casting `BLImageCore` to `BLImage`, but that's after the image has already been loaded.

**Reply by f-saez** [reply]
In the fiddle, it works. Obviously, may I say :-))

To be more precise, I'm writing bindings for Mojo and I've wrote the same thing for Rust years ago (but I've keep it more or less private). I've run in the same trouble with Rust and I've "solved" it by using libjpeg.

By the way, what does 2 means ?

I'm sure it is something on my side. I've probably replicate my first misunderstanding.

**Reply by kobalicek** [reply]
`blImageReadFromFile` requires a constructed image (by `blImageInit`), so maybe forgetting to init the instance is the problem?

Otherwise it looks like there is a mistake in the bindings, which is something hard to diagnose from my side. Error `2` is suspicious. In some rare situations Blend2D may fail to translate an error and pass a libC error in such case (like ENOMEM), but I think that's just unlikely to happen as it should basically always translate from libC error to a corresponding Blend2D error.

I think you really need to step in with a debugger and to see where it fails. I'm interested in seeing the place.

**Comment by f-saez** [comment]
For what it's worth, here is my code in Mojo : 
```
var filename = String("file.png") # a zero-terminated string, like in C
var core = BLImageCore()  
var res = b2d._handle.get_function[blImageInit]("blImageInit")(UnsafePointer(core))
if res==BL_SUCCESS:
    var ptr0 = UnsafePointer[BLImageCore](core)  # 
    var ptr1 = filename.__str__().unsafe_uint8_ptr() # 
    var ptr2 = UnsafePointer[UInt8]() # null ptr, I guess. It works for other functions, at least
    var res = self._b2d._handle.get_function[blImageReadFromFile]("blImageReadFromFile")(ptr0, ptr1, ptr2)
```

I'll keep you updated. :-)

**Comment by f-saez** [comment]
It seems to be a problem with the filename or the reading of the file.
I've traced (the hard way as I'm no C++ coder) the problem in filesystem.cpp :
```
BL_API_IMPL BLResult blFileOpen(BLFileCore* self, const char* fileName, BLFileOpenFlags openFlags) noexcept {
...
  // NOTE: Do not close the file before calling `open()`. We should behave atomically, which means that we won't
  // close the existing file if `open()` fails...
  int fd = BL_FILE64_API(open)(fileName, of, om);
```
fd==-1 means the file can't be read, but I can't understand why.
just in case : of = 0 and om=438

It happens whatever filename I use.

It also means that I should not be able to reproduce the bug by using readFromData.

**Comment by kobalicek** [comment]
Okay, I have fixed an error translation (it was a bug), but still, the (2) error is ENOENT, which means the file doesn't exist. So most likely the path is wrong in this case. I'm not sure how MOJO works, but Blend2D needs a null terminated string in this case.

**Comment by f-saez** [comment]
Mojo strings are just C Strings.
I've used them for the text part (GlyphBuffer, ...)of Blend2D without any trouble.

**Comment by f-saez** [comment, marked as answer]
Ok, I may have found the solution.
Mojo, as Rust, has got a lifetime management feature and it is the culprit.

Let me explain :
```
var filename = Path("image.jpg")
```
filename is a Path object
To get the pointer of the text of this object, I need to do this: 
```
var ptr = filename.__str__().unsafe_uint8_ptr()
```
If I checked ptr in Mojo, I've got: 0x563c47f96020 and I've got the same value in blImageReadFromFile.
If I print fileName in Mojo, I've got "image.jpg" but in blImageReadFromFile I've got garbage.

The devil is in the details, as always.

It means that  filename.__str__().unsafe_uint8_ptr() doesn't give me a pointer on the text of the Path.
It means that  filename.__str__().unsafe_uint8_ptr() create a temporary variable that contains the text of the path, give me the pointer of this variable and immediately, and silently, discards the variable.
I haven't a way to specify the lifetime of variables and Mojo discard a variable as soon as it can, but in the case of FFI, it's often a problem.

I've already asked how to specify a lifetime on Mojo's Discord three weeks ago but no one has ever answer, so I'll try to post the same question on their github.

**Comment by f-saez** [comment]
just in case you got time to waste, here is the link to the first shot of the bindings :

https://github.com/f-saez/blend2d-mojo/tree/main

thanks again for your time.
